p2p/src/peer_manager/peerdb/address_tables/table.rs

```diff
@@ -0,0 +1,532 @@
+    /// directly. Instead, addresses are stored in a map indexed by a monotonously increasing
+    /// ids and the array only contains the ids.
+    // TODO: use something more efficient than a vector of vectors.
+    buckets: Vec<Vec<EntryId>>,
+    /// The addresses map.
+    addresses: BTreeMap<EntryId, SocketAddress>,
+    /// The maximum value of an id plus one. This value will be stored in `buckets` to indicate
+    /// that the entry is unoccupied. Normally, it will be equal to EntryId::MAX, but we allow
+    /// to override it for testing.
+    id_max: EntryId,
+    /// Arbitrary value; this is used as an additional "key" to randomize bucket selection.
+    random_key: RandomKey,
+}
+
+impl Table {
+    pub fn new(bucket_count: usize, bucket_size: usize, random_key: RandomKey) -> Self {
+        Self::new_generic(bucket_count, bucket_size, random_key, EntryId::MAX)
+    }
+
+    fn new_generic(
+        bucket_count: usize,
+        bucket_size: usize,
+        random_key: RandomKey,
+        id_max: EntryId,
+    ) -> Self {
+        assert!(id_max as usize >= bucket_count * bucket_size);
+
+        let mut buckets = Vec::with_capacity(bucket_count);
+
+        for _ in 0..bucket_count {
+            let mut bucket = Vec::with_capacity(bucket_size);
+            bucket.resize(bucket_size, id_max);
+            buckets.push(bucket);
+        }
+
+        Self {
+            buckets,
+            addresses: BTreeMap::new(),
+            random_key,
+            id_max,
+        }
+    }
+
+    // TODO: in bitcoin the "new" table bucket index is calculated differently from
+    // the "tried" one:
+    // 1) They also hash the "source", i.e. the address of the peer or dns seed
+    // that advertised this address. This results in the same address being added
+    // to the table multiple times if it was advertised by multiple sources, so
+    // addresses that are advertised more have more chances to stay in the table.
+    // 2) Only the "addr group" part of the address is included in the hash, so all
+    // addresses from the same addr group will end up in the same bucket.
+    // Should we do something similar?
+    fn bucket_idx(&self, addr: &SocketAddress) -> usize {
+        let addr_hash = calc_hash!(self.random_key, addr);
+        let addr_group = AddressGroup::from_peer_address(&addr.as_peer_address());
+
+        // Note: addresses from a certain address group can be spread over at most BUCKETS_PER_GROUP
+        // buckets.
+        let final_hash = calc_hash!(self.random_key, addr_group, addr_hash % BUCKETS_PER_GROUP);
+
+        (final_hash % self.buckets.len() as u64) as usize
+    }
+
+    fn bucket_pos(&self, addr: &SocketAddress, bucket_idx: usize) -> usize {
+        let hash = calc_hash!(self.random_key, addr, bucket_idx);
+        (hash % self.buckets[bucket_idx].len() as u64) as usize
+    }
+
+    pub fn entry(&self, addr: &SocketAddress) -> Option<&SocketAddress> {
```

Comment:
IMO `get` wouldn't be a good name, because the function may return a different address. So it's not like `BTreeSet::get`. And `get_entry` seems redundant.
Also, I already use the word "entry" in other places to refer to an "item" of the table. So IMO this function must have "entry" in its name. I don't mind using some other word, to avoid confusion with `BTreeMap::entry`, but I'm not sure which one would be descriptive enough.